src/MySqlConnector/Serialization/MySqlSession.cs

```diff
@@ -35,16 +36,97 @@ public MySqlSession(ConnectionPool pool, int poolGeneration)
 		public ConnectionPool Pool { get; }
 		public int PoolGeneration { get; }
 		public string DatabaseOverride { get; set; }
+		public IPAddress IPAddress => (m_tcpClient?.Client.RemoteEndPoint as IPEndPoint)?.Address;
 
 		public void ReturnToPool() => Pool?.Return(this);
 
-		public bool IsConnected => m_state == State.Connected;
+		public bool IsConnected
+		{
+			get
+			{
+				lock (m_lock)
+					return m_state == State.Connected;
+			}
+		}
+
+		public bool TryStartCancel(MySqlCommand command)
+		{
+			lock (m_lock)
+			{
+				if (m_activeCommand != command)
+					return false;
+				VerifyState(State.Querying, State.CancelingQuery);
+				if (m_state != State.Querying)
+					return false;
+				m_state = State.CancelingQuery;
+			}
+
+			return true;
+		}
+
+		public void DoCancel(MySqlCommand commandToCancel, MySqlCommand killCommand)
+		{
+			lock (m_lock)
+			{
+				if (m_activeCommand != commandToCancel)
+					return;
+
+				try
+				{
+					killCommand.ExecuteNonQuery();
```

Comment:
I wasn't a fan of this, but I wanted to avoid the following problem:

thread 1 | thread 2
--- | ---
cmd1.Execute | 
_ | cmd1.Cancel
cmd1 finishes | open new connection
cmd2.Execute | wait for new connection to open
_ | KILL QUERY x
_ | ^ cmd2 is killed

We could dial down the `CommandTimeout` (once that is implemented #67) but I couldn't think of another approach here that makes sure cancellation (which has to run asynchronously) actually cancels the right command instead of just killing what happens to be actively running on that connection.